File: sword_offer/python/61_序列化二叉树.py
#!/bin/bash/ python3
#-*- coding:utf-8 -*-

import time
import TreeNode

"""
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
"""

class Solution:
    def Serialize(self, root):
        """
        序列化
        """
        if root == None:
            return "#,"
        else:
            return str(root.val) + "," + self.Serialize(root.left) + self.Serialize(root.right)
    
    def Deserialize(self, s):
        """
        反序列化
        """
        def Des(lst):
            if lst == []:
                return None
            val = lst.pop(0)
            if val == "#":
                return None
            else:
                root = TreeNode.TreeNode(int(val))
                root.left = Des(lst)
                root.right = Des(lst)
                return root
        
        lst = s.split(",") # 利用,分割str为list
        return Des(lst)
    
    def test(self, vec):
        """
        测试函数
        """
        head = TreeNode.TreeNode()
        ori = head.creatFromFrontOrder(vec)
        print("=====")
        start = time.time()
        s = self.Serialize(ori)
        rec = self.Deserialize(s)
        end = time.time()
        
        print("time(us): {:>5.2f}, str: {}".format((end - start)*10**6, s))
        print("ori-front-order: ", end="")
        head.printFrontOrder(ori);
        print("\nrec-front-order: ", end="")
        head.printFrontOrder(rec)
        print()


def main():
    vec = [1, 2, -1, -1, 3, 4, -1, -1, 5, -1, -1]

    s = Solution()
    s.test(vec)


if __name__ == "__main__":
    main()

File: sword_offer/cpp/61_序列化二叉树.cpp
#include <string>
#include <vector>
#include <stdio.h>
#include <sys/time.h>

#include "TreeNode.h"

/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
*/

class Solution {
private:
    std::string ret;

public:
    // 1. string接口
    // 序列化，先序遍历
    // 使用逗号分割，空节点为#
    std::string Serialize(TreeNode *root) {
        if (root == nullptr) {
            return "#,";
        } else {
            return std::to_string(root->val) + "," + Serialize(root->left) + Serialize(root->right);
        }
    }
    
    // 反序列化
    TreeNode* Deserialize(std::string str) {
        int index = 0;
        return Des(str, index);
    }

    // 递归函数
    TreeNode* Des(std::string& str, int& i) {
        if (str.empty() || i >= str.size()) { // 防止溢出
            return nullptr;
        }

        if (str[i] == '#') { // 空节点
            i += 2;
            return nullptr;
        }

        // 提取当前结点值
        int curr_val = 0;
        for (; str[i] != ','; i++) {
            curr_val = curr_val * 10 + (str[i] - '0');
        }
        i++; // 跳过标志符号','
        
        TreeNode* root = new TreeNode(curr_val);
        root->left = Des(str, i);
        root->right = Des(str, i);
        return root;
    }
    
    // 2. char*接口（用于牛客网提交，本地不进行调试）
    char* char_Serialize(TreeNode *root) {    
        if (root == nullptr) {
            return nullptr;
        }
        this->ret = this->Serialize(root);
        return (char*)(ret.c_str());
    }
    
    TreeNode* char_Deserialize(char *str) {
        if (str == nullptr) {
            return nullptr;
        }
        std::string tmp = str;
        return Deserialize(tmp);
    }

    // 测试函数
    void test(std::vector<int>& vec) {
        struct timeval start, end;
        std::string str;
        TreeNode* rec = nullptr;
        TreeNode* ori = TreeNode_creatFromFrontOrder(vec);

        printf("=====\n");
        gettimeofday(&start, nullptr);
        str = this->Serialize(ori);
        rec = this->Deserialize(str);
        gettimeofday(&end, nullptr);
        
        printf("times(us): %ld, str: %s\n", end.tv_usec - start.tv_usec, str.c_str());
        printf("ori-front-order: ");
        TreeNode_printFrontOrder(ori);
        printf("\nrec-front-order: ");
        TreeNode_printFrontOrder(rec);
        printf("\n");
    }
};

int main(int argc, char* argv[])
{
    std::vector<int> vec = {1, 2, -1, -1, 3, 4, -1, -1, 5, -1, -1};

    Solution s;
    s.test(vec);

    return 0;
}